File: go/creational/builder.go
package main

import "fmt"

// product
type Report struct {
	title   string
	header  string
	content string
	footer  string
}

func (r *Report) Display() {
	fmt.Println("Report:")
	fmt.Println("Header: ", r.header)
	fmt.Println("Title: ", r.title)
	fmt.Println("Content: ", r.content)
	fmt.Println("Footer: ", r.footer)
}

// builder interface
type ReportBuilder interface {
	SetHeader(header string) ReportBuilder
	SetTitle(title string) ReportBuilder
	SetContent(content string) ReportBuilder
	SetFooter(footer string) ReportBuilder
	Build() *Report
}

// concerete builders
type PDFReportBuilder struct {
	report *Report
}

func NewPDFReportBuilder() *PDFReportBuilder {
	return &PDFReportBuilder{
		report: &Report{},
	}
}

func (pr *PDFReportBuilder) SetTitle(title string) ReportBuilder {
	pr.report.title = title
	return pr
}

func (pr *PDFReportBuilder) SetHeader(header string) ReportBuilder {
	pr.report.header = header
	return pr
}

func (pr *PDFReportBuilder) SetContent(content string) ReportBuilder {
	pr.report.content = content
	return pr
}

func (pr *PDFReportBuilder) SetFooter(footer string) ReportBuilder {
	pr.report.footer = footer
	return pr
}

func (pr *PDFReportBuilder) Build() *Report {
	return pr.report
}

type ExceleportBuilder struct {
	report *Report
}

func NewExcelReportBuilder() *ExceleportBuilder {
	return &ExceleportBuilder{
		report: &Report{},
	}
}

func (pr *ExceleportBuilder) SetTitle(title string) ReportBuilder {
	pr.report.title = title
	return pr
}

func (pr *ExceleportBuilder) SetHeader(header string) ReportBuilder {
	pr.report.header = header
	return pr
}

func (pr *ExceleportBuilder) SetContent(content string) ReportBuilder {
	pr.report.content = content
	return pr
}

func (pr *ExceleportBuilder) SetFooter(footer string) ReportBuilder {
	pr.report.footer = footer
	return pr
}

func (pr *ExceleportBuilder) Build() *Report {
	return pr.report
}

func main() {

	pdfBuilder := NewPDFReportBuilder()
	pdfReport := pdfBuilder.
		SetTitle("learning desing pattern").
		SetHeader("Builder pattern").
		SetContent("lorem ipsum..").
		SetFooter("desing pattern cookbook").
		Build()
	pdfReport.Display()

	excelBuilder := NewExcelReportBuilder()
	excelReport := excelBuilder.
		SetTitle("learning desing pattern").
		SetHeader("Builder pattern").
		SetContent("lorem ipsum..").
		SetFooter("desing pattern cookbook").
		Build()
	excelReport.Display()
}

File: python/creational/builder.py
from abc import ABC, abstractmethod

# product
class Report:
    def __init__(self):
        self.title = None
        self.header = None
        self.content = None
        self.footer = None

    def display(self):
        print("Report:")
        print(f" Title: {self.title}")
        print(f" Header: {self.header}")
        print(f" Content: {self.content}")
        print(f" Footer: {self.footer}")

# builder interface
class ReportBuilder(ABC):
    
    @abstractmethod
    def set_header(self, header):
        pass

    @abstractmethod
    def set_title(self, tile):
        pass

    @abstractmethod
    def set_content(self, content):
        pass

    @abstractmethod
    def set_footer(self, footer):
        pass

    @abstractmethod
    def build(self):
        pass

# concerete builders

class PDFReportBuilder(ReportBuilder):
    def __init__(self):
        self.report = Report()

    def set_header(self, header):
        self.report.header = header
        return self
    
    def set_content(self, content):
        self.report.content = content
        return self

    def set_title(self, title):
        self.report.title = title
        return self

    def set_footer(self, footer):
        self.report.footer = footer
        return self

    def build(self) -> Report:
        return self.report
    
class CSVBuilder(ReportBuilder):
    def __init__(self):
        self.report = Report()
    def set_header(self, header):
        self.report.header = header
    
    def set_content(self, content):
        self.report.content = content
        return self

    def set_title(self, title):
        self.report.title = title
        return self

    def set_footer(self, footer):
        self.report.footer = footer
        return self

    def build(self) -> Report:
        return self.report

if __name__ == "__main__":
    
    pdf_builder = PDFReportBuilder()
    pdf_report = (
        pdf_builder
        .set_title("learning desing pattern")
        .set_header("Builder pattern")
        .set_content("lorem ipsum...")
        .set_footer("desing pattern cookbook")
        .build()
    )
    pdf_report.display()
    